File: packages/pushduck-go/pushduck/sign.go
package pushduck

// SigV4 query presigning.
//
// This is the only cryptography in the request path, and the only place where
// a Go implementation can be self-consistently wrong: a signature derived with
// the wrong canonical request is still a valid-looking URL, and only the
// provider rejects it. The TypeScript implementation delegates to aws4fetch;
// here it is written out, because pulling in the AWS SDK for one signature
// would dwarf the rest of this package.
//
// Reference: "Authenticating Requests: Using Query Parameters (AWS Signature
// Version 4)" — the canonical request, the string to sign, and the signing key
// derivation are all fixed by that document and none of it is negotiable.

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/url"
	"sort"
	"strings"
	"time"
)

const (
	algorithm       = "AWS4-HMAC-SHA256"
	unsignedPayload = "UNSIGNED-PAYLOAD"
)

func hmacSHA256(key []byte, data string) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(data))
	return mac.Sum(nil)
}

func sha256Hex(data string) string {
	sum := sha256.Sum256([]byte(data))
	return hex.EncodeToString(sum[:])
}

// uriEncode percent-encodes per S3's rules rather than Go's.
//
// `url.QueryEscape` encodes a space as `+`, which S3 does not accept in a
// canonical request, and leaves some characters unescaped that S3 expects
// escaped. Getting this wrong produces a signature mismatch on exactly the
// filenames users complain about — the ones with spaces and non-ASCII
// characters.
func uriEncode(s string, encodeSlash bool) string {
	var b strings.Builder

	for _, c := range []byte(s) {
		switch {
		case (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
			(c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~':
			b.WriteByte(c)
		case c == '/':
			if encodeSlash {
				b.WriteString("%2F")
			} else {
				b.WriteByte('/')
			}
		default:
			b.WriteString(fmt.Sprintf("%%%02X", c))
		}
	}

	return b.String()
}

// signingKey derives the date/region/service-scoped key.
func signingKey(secret, dateStamp, region, service string) []byte {
	kDate := hmacSHA256([]byte("AWS4"+secret), dateStamp)
	kRegion := hmacSHA256(kDate, region)
	kService := hmacSHA256(kRegion, service)
	return hmacSHA256(kService, "aws4_request")
}

// presignOptions is everything a signature depends on.
type presignOptions struct {
	Method    string
	Host      string
	Path      string
	Region    string
	Query     url.Values
	Headers   map[string]string
	ExpiresIn int
	Now       time.Time
}

// presign returns a fully signed URL.
//
// Every signed header must also be sent by the client, which is why the caller
// gets `requiredHeaders` back rather than being expected to guess.
func (c *Config) presign(opts presignOptions) string {
	now := opts.Now.UTC()
	amzDate := now.Format("20060102T150405Z")
	dateStamp := now.Format("20060102")

	credentialScope := strings.Join(
		[]string{dateStamp, opts.Region, "s3", "aws4_request"}, "/",
	)

	// Signed headers must be lowercase and sorted; the canonical form is what
	// is hashed, not what is sent.
	signedHeaderNames := make([]string, 0, len(opts.Headers)+1)
	canonicalHeaders := map[string]string{"host": opts.Host}
	signedHeaderNames = append(signedHeaderNames, "host")

	for name, value := range opts.Headers {
		lower := strings.ToLower(name)
		canonicalHeaders[lower] = strings.TrimSpace(value)
		signedHeaderNames = append(signedHeaderNames, lower)
	}
	sort.Strings(signedHeaderNames)

	query := url.Values{}
	for key, values := range opts.Query {
		query[key] = values
	}
	query.Set("X-Amz-Algorithm", algorithm)
	query.Set("X-Amz-Credential", c.AccessKeyID+"/"+credentialScope)
	query.Set("X-Amz-Date", amzDate)
	query.Set("X-Amz-Expires", fmt.Sprintf("%d", opts.ExpiresIn))
	query.Set("X-Amz-SignedHeaders", strings.Join(signedHeaderNames, ";"))

	// Temporary credentials carry a third part, and omitting it makes every
	// STS-issued credential fail with a signature error that names the
	// signature rather than the missing token.
	if c.SessionToken != "" {
		query.Set("X-Amz-Security-Token", c.SessionToken)
	}

	// The canonical query string is sorted by encoded key, with encoded values.
	keys := make([]string, 0, len(query))
	for key := range query {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	pairs := make([]string, 0, len(keys))
	for _, key := range keys {
		pairs = append(pairs, uriEncode(key, true)+"="+uriEncode(query.Get(key), true))
	}
	canonicalQuery := strings.Join(pairs, "&")

	headerLines := make([]string, 0, len(signedHeaderNames))
	for _, name := range signedHeaderNames {
		headerLines = append(headerLines, name+":"+canonicalHeaders[name]+"\n")
	}

	canonicalRequest := strings.Join([]string{
		opts.Method,
		uriEncode(opts.Path, false),
		canonicalQuery,
		strings.Join(headerLines, ""),
		strings.Join(signedHeaderNames, ";"),
		unsignedPayload,
	}, "\n")

	stringToSign := strings.Join([]string{
		algorithm,
		amzDate,
		credentialScope,
		sha256Hex(canonicalRequest),
	}, "\n")

	signature := hex.EncodeToString(
		hmacSHA256(signingKey(c.SecretAccessKey, dateStamp, opts.Region, "s3"), stringToSign),
	)

	return "https://" + opts.Host + uriEncode(opts.Path, false) +
		"?" + canonicalQuery + "&X-Amz-Signature=" + signature
}

File: packages/pushduck-py/pushduck/sign.py
"""SigV4 query presigning.

The only cryptography in the request path, and the only place a Python
implementation can be self-consistently wrong: a signature derived from the
wrong canonical request is still a valid-looking URL, and only the provider
rejects it.

Written out rather than delegated to boto3. One signature does not justify
dragging botocore into every application that installs this package, and the
canonical-request rules are fixed by the SigV4 specification — there is nothing
to keep up with.
"""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone
from typing import Dict, Mapping, Optional

_ALGORITHM = "AWS4-HMAC-SHA256"
_UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD"

# Characters S3 leaves unescaped in a canonical request. Deliberately not
# `urllib.parse.quote`'s default set: that leaves `/` alone in query values and
# encodes a space as `%20` only sometimes, and either difference produces a
# signature mismatch on exactly the filenames users complain about.
_UNRESERVED = frozenset(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ" "abcdefghijklmnopqrstuvwxyz" "0123456789" "-_.~"
)


def uri_encode(value: str, encode_slash: bool = True) -> str:
    """Percent-encode per S3's rules."""
    out = []
    for byte in value.encode("utf-8"):
        char = chr(byte)
        if char in _UNRESERVED:
            out.append(char)
        elif char == "/":
            out.append("%2F" if encode_slash else "/")
        else:
            out.append(f"%{byte:02X}")
    return "".join(out)


def _hmac(key: bytes, message: str) -> bytes:
    return hmac.new(key, message.encode("utf-8"), hashlib.sha256).digest()


def _sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def _signing_key(secret: str, date_stamp: str, region: str, service: str) -> bytes:
    date = _hmac(f"AWS4{secret}".encode("utf-8"), date_stamp)
    scoped_region = _hmac(date, region)
    scoped_service = _hmac(scoped_region, service)
    return _hmac(scoped_service, "aws4_request")


def presign(
    *,
    access_key_id: str,
    secret_access_key: str,
    session_token: Optional[str],
    method: str,
    scheme: str,
    host: str,
    path: str,
    region: str,
    query: Optional[Mapping[str, str]] = None,
    headers: Optional[Mapping[str, str]] = None,
    expires_in: int = 3600,
    now: Optional[datetime] = None,
) -> str:
    """Return a fully signed URL.

    Every signed header must also be sent by the client, which is why callers
    receive ``required_headers`` back rather than being expected to guess.
    """
    moment = (now or datetime.now(timezone.utc)).astimezone(timezone.utc)
    amz_date = moment.strftime("%Y%m%dT%H%M%SZ")
    date_stamp = moment.strftime("%Y%m%d")

    credential_scope = f"{date_stamp}/{region}/s3/aws4_request"

    # Signed headers are lowercase and sorted; the canonical form is what gets
    # hashed, not what is sent.
    canonical_headers: Dict[str, str] = {"host": host}
    for name, value in (headers or {}).items():
        canonical_headers[name.lower()] = value.strip()
    signed_header_names = sorted(canonical_headers)

    parameters: Dict[str, str] = dict(query or {})
    parameters.update(
        {
            "X-Amz-Algorithm": _ALGORITHM,
            "X-Amz-Credential": f"{access_key_id}/{credential_scope}",
            "X-Amz-Date": amz_date,
            "X-Amz-Expires": str(expires_in),
            "X-Amz-SignedHeaders": ";".join(signed_header_names),
        }
    )

    # Temporary credentials carry a third part. Omitting it makes every
    # STS-issued credential — ECS task roles, EKS IRSA, OIDC AssumeRole — fail
    # with an error that names the signature rather than the missing token.
    if session_token:
        parameters["X-Amz-Security-Token"] = session_token

    canonical_query = "&".join(
        f"{uri_encode(key)}={uri_encode(parameters[key])}" for key in sorted(parameters)
    )

    canonical_request = "\n".join(
        [
            method,
            uri_encode(path, encode_slash=False),
            canonical_query,
            "".join(f"{name}:{canonical_headers[name]}\n" for name in signed_header_names),
            ";".join(signed_header_names),
            _UNSIGNED_PAYLOAD,
        ]
    )

    string_to_sign = "\n".join(
        [_ALGORITHM, amz_date, credential_scope, _sha256_hex(canonical_request)]
    )

    signature = hmac.new(
        _signing_key(secret_access_key, date_stamp, region, "s3"),
        string_to_sign.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return (
        f"{scheme}://{host}{uri_encode(path, encode_slash=False)}"
        f"?{canonical_query}&X-Amz-Signature={signature}"
    )
